refactor: move from bevy_heron to bevy_rapier2d

CashWasabi opened the pull request:
## Summary
The refactor consists of basically two things:
- show use of `bevy_rapier2d` in platformer example
- add `bevy_rapier2d` in cargo file and remove `heron`

For the most part I tried to not change too much of the platformer example. 
There we're some minor things though that came up whith which I had to struggle a bit.

## Notable Details
- rename `CollisionShape` -> `Collider`
- rename `Velocity.linear` -> `Velocity.linvel`
- create `SensorBundle` since there is no `RigidBody::Sensor` in `rapier2d`
- set player friction to `0.0` because the collider kept sticking to the wall when moving against it in air
- add `RapierDebugRenderPlugin` to showcase colliders with gizmos
- `pause_physics_during_load` uses RapierConfiguration now
- add `ExternalForce` to `ColliderBundle` for jumps

## Issues
I noticed that my fix for `pause_physics_during_load` might not work as excepted. Sometimes when building I have to press `R` to reload the level when I want to have the `GroundSensor` working. This might be a loading issue but I didn't find the correct solution yet.

Trouv commented on examples/platformer/systems.rs:
```diff
@@ -21,12 +21,12 @@ pub fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
 
 pub fn pause_physics_during_load(
```
I think we may be able to get rid of this system now, actually. The issue it addressed with Heron doesn't seem to be a problem anymore.

CashWasabi replied:
Deleted it.

Trouv commented on examples/platformer/systems.rs:
```diff
@@ -46,31 +46,42 @@ pub fn dbg_player_items(
 
         if input.just_pressed(KeyCode::Space) && (ground_detection.on_ground || climber.climbing) {
-            velocity.linear.y = 450.;
+            force.y = 500.;
```
I think we probably need to increase this. On my machine, the jump is tiny, I can't even get over the skulls.

CashWasabi replied:
I changed it to be 500 and it's using `velocity.linvel` directly know. This might fix the inconsistent behavior.

Trouv commented on examples/platformer/systems.rs:
```diff
@@ -46,31 +46,42 @@ pub fn dbg_player_items(
         }
 
         if input.just_pressed(KeyCode::Space) && (ground_detection.on_ground || climber.climbing) {
-            velocity.linear.y = 450.;
```
Is there a reason that we're applying an external force to it instead of a velocity?

CashWasabi replied:
Not really, I got rid of it and also got rid of the `ExternalForce` component in the query.

Trouv commented on examples/platformer/systems.rs:
```diff
@@ -449,22 +441,22 @@ pub fn update_level_selection(
-            };
+    for (entity, shape, transform) in &detect_ground_for {
+        if let Some(cuboid) = shape.as_cuboid() {
+            let Vec2 { x: hx, y: hy } = cuboid.half_extents();
```
I'd prefer we keep variable names verbose if possible. `half_extents_x` `half_extents_y`

Trouv commented on examples/platformer/systems.rs:
```diff
@@ -449,22 +441,22 @@ pub fn update_level_selection(
-            let sensor_translation = Vec3::new(0., -half_extends.y, 0.) / transform.scale;
+            let detector_shape = Collider::cuboid(hx / 2.0, 2.);
+
+            let sensor_translation = Vec3::new(0., -hy, 0.) / transform.scale;
```
So this ` / transform.scale` thing was a bit of a workaround for an old heron issue. I think we might be able to get rid of it now. It might be causing the ground sensor to be way lower than it should be currently

CashWasabi replied:
Removed it. :)

Trouv commented on examples/platformer/systems.rs:
```diff
@@ -480,31 +472,24 @@ pub fn ground_detection(
-                    Ok(RigidBody::Sensor) => {
-                        // don't consider sensors to be "the ground"
-                    }
+                CollisionEvent::Started(a, b, _event_flag) => match colliders.get(*b) {
```
We may have to do something similar here that we do for the ladders where we try both orderings of `a` and `b`. By that I mean, first treat `a` as the sensor and `b` as the ground, and if it doesn't match then try `a` as the ground and `b` as the sensor. This is because collision events are only being sent once with a particular order per collision with bevy_rapier2d instead of twice with different orderings of `a` and `b`. I had to do this for my ludum dare game, which happened to be a platformer using `bevy_rapier2d`: https://github.com/Trouv/ludum-dare-51/blob/main/src/player.rs#L296

CashWasabi replied:
Great code base! I used it to refactor the `ground_detection` and `detect_climb_range` function.

Trouv commented on examples/platformer/systems.rs:
```diff
@@ -46,31 +46,42 @@ pub fn dbg_player_items(
             let down = if input.pressed(KeyCode::S) { 1. } else { 0. };
 
-            velocity.linear.y = (up - down) * 200.;
+            velocity.linvel.y = (up - down) * 200.;
```
There seems to be some bug with players not staying on ladders when neither button is pressed like they used to. I imagine that's why you changed the `just_pressed` to `pressed` for climbing inputs. I'd like to fix this if possible

Trouv reviewed (changes requested):
Thanks so much for the contribution! I should have mentioned that I actually made a platformer with rapier for ludum dare 51, and it may be a half-decent resource for sorting out some of the issues bc the platforming turned out pretty stable there: https://github.com/Trouv/ludum-dare-51

CashWasabi commented on examples/platformer/components.rs:
```diff
@@ -57,21 +60,26 @@ impl From<EntityInstance> for ColliderBundle {
+    fn from(_: IntGridCell) -> ColliderBundle {
+        ColliderBundle::default()
+    }
+}
```
Deleted it.

Trouv commented on examples/platformer/components.rs:
```diff
@@ -3,75 +3,79 @@ use bevy_ecs_ldtk::{prelude::*, utils::ldtk_pixel_coords_to_translation_pivoted}
 pub struct ColliderBundle {
-    pub collider: CollisionShape,
+    pub collider: Collider,
     pub rigid_body: RigidBody,
     pub velocity: Velocity,
-    pub rotation_constraints: RotationConstraints,
-    pub physic_material: PhysicMaterial,
+    pub force: ExternalForce,
+    pub rotation_constraints: LockedAxes,
+    pub gravity_scale: GravityScale,
+    pub restitution: Restitution,
+    pub friction: Friction,
+    pub density: ColliderMassProperties,
+}
```
I think now we can get rid of `ExternalForce` and `Restitution` no?

CashWasabi replied:
That's true. I kinda forgot to get rid of the unnecessary leftovers.

Trouv commented on examples/platformer/systems.rs:
```diff
@@ -258,76 +240,62 @@ pub fn spawn_wall_collision(
 pub fn detect_climb_range(
     mut climbers: Query<&mut Climber>,
-    climbables: Query<&Climbable>,
+    climbables: Query<Entity, (With<Collider>, With<Sensor>, With<Climbable>, Without<RigidBody>)>,
```
I think `With<Climbable>` would suffice right? Or am I missing something

CashWasabi replied:
Yes, `With<Climbable>` is enough already and looks nicer, too.

Trouv reviewed (changes requested):
This works great, just a couple minor changes left.

Also - please run `cargo fmt`, I know the CI will complain about it after I approve this.

CashWasabi commented:
I added the changes you last mentioned and ran `cargo fmt` on the codebase. I also squashed the commit since the last two commits we're just small fixes with no real information.

Trouv commented:
> I also squashed the commit since the last two commits we're just small fixes with no real information.

In the future, there's no need to do this for this repo since I squash commits on merge (as of 0.4) anyway. Actually, having the smaller commits before merge makes it easier to see what has changed since the last review

janos-r commented:
I was working on the same update (not noticing this PR exists). In my case, I found it necessary to use `.normalize_or_zero()` instead of `.normalize()` because the first mob (spawning closest to the player), when returning, got to the exact same spot from where he was starting, creating his vector zero, having .normalize panic. I wonder why It doesn't happen here (that he doesn't return to the exact spot).
